File: implementations/ts/src/schema.ts
// HyperSchemas and the schema registry (SPEC-3 §2-3 §6, ERRATA-2 E10/E13). The registry indexes
// schemas by name AND by term hash; pinned refs resolve by hash and are immutable by construction.
// Since issue #23 it also holds resolution Schemas ("readings"), indexed the same two ways, so an
// expand term can name both halves of a child's lens and be validated at build time.

import type { SchemaRefT, Term } from "./eval.js";
import type { Schema } from "./resolution.js";
import { schemaHash, termHash } from "./term-io.js";

export interface HyperSchema {
  readonly name: string;
  readonly alg: number; // L2 algebra version
  readonly body: Term; // an HView-sort term, a function of the ambient root
}

// refs are derived from the body — every expand/fix schema reference (E10).
export function collectRefs(term: Term): SchemaRefT[] {
  const out: SchemaRefT[] = [];
  const walk = (t: Term): void => {
    switch (t.kind) {
      case "input":
        return;
      case "select":
      case "mask":
      case "group":
      case "prune":
      case "resolve":
        walk(t.of);
        return;
      case "union":
      case "intersect":
        walk(t.left);
        walk(t.right);
        return;
      case "difference":
        walk(t.of);
        walk(t.without);
        return;
      case "expand":
        out.push(t.schema);
        walk(t.of);
        return;
      case "fix":
        out.push(t.schema);
        return;
    }
  };
  walk(term);
  return out;
}

// Reading refs are derived the same way — every expand's `reading` (issue #23). Kept separate
// from collectRefs because they resolve against a different index (readings, not hyperschemas).
export function collectReadingRefs(term: Term): SchemaRefT[] {
  const out: SchemaRefT[] = [];
  const walk = (t: Term): void => {
    switch (t.kind) {
      case "input":
      case "fix":
        return;
      case "select":
      case "mask":
      case "group":
      case "prune":
      case "resolve":
        walk(t.of);
        return;
      case "union":
      case "intersect":
        walk(t.left);
        walk(t.right);
        return;
      case "difference":
        walk(t.of);
        walk(t.without);
        return;
      case "expand":
        if (t.reading !== undefined) out.push(t.reading);
        walk(t.of);
        return;
    }
  };
  walk(term);
  return out;
}

export class SchemaRegistry {
  private constructor(
    private readonly byName: ReadonlyMap<string, HyperSchema>,
    private readonly byHash: ReadonlyMap<string, HyperSchema>,
    private readonly readingsByName: ReadonlyMap<string, Schema>,
    private readonly readingsByHash: ReadonlyMap<string, Schema>,
  ) {}

  // Rejects duplicate names, unresolved refs (gather AND reading), and reference cycles
  // (SPEC-3 §3). Data cycles remain legal — the DAG constraint is on programs, not data.
  static build(schemas: readonly HyperSchema[], readings: readonly Schema[] = []): SchemaRegistry {
    const readingsByName = new Map<string, Schema>();
    const readingsByHash = new Map<string, Schema>();
    for (const r of readings) {
      if (r.name === undefined) {
        throw new Error("a registered reading must carry a name (issue #23)");
      }
      if (readingsByName.has(r.name)) throw new Error(`duplicate reading name: ${r.name}`);
      readingsByName.set(r.name, r);
      const h = schemaHash(r);
      // As with hyperschema bodies, two names MAY share a hash; first registration wins.
      if (!readingsByHash.has(h)) readingsByHash.set(h, r);
    }
    const resolveReadingRef = (ref: SchemaRefT, from: string): void => {
      const found =
        ref.kind === "name" ? readingsByName.get(ref.name) : readingsByHash.get(ref.hash);
      if (found === undefined) {
        const label = ref.kind === "name" ? ref.name : `pinned:${ref.hash.slice(0, 12)}…`;
        throw new Error(`schema ${from} references unknown reading ${label} (issue #23)`);
      }
    };
    const byName = new Map<string, HyperSchema>();
    const byHash = new Map<string, HyperSchema>();
    const hashOf = new Map<string, string>(); // name -> term hash
    for (const s of schemas) {
      if (byName.has(s.name)) throw new Error(`duplicate schema name: ${s.name}`);
      byName.set(s.name, s);
      const h = termHash(s.body);
      hashOf.set(s.name, h);
      // Two names MAY share a body hash; first registration wins the hash index.
      if (!byHash.has(h)) byHash.set(h, s);
      for (const r of collectReadingRefs(s.body)) resolveReadingRef(r, s.name);
    }
    const resolveName = (ref: SchemaRefT, from: string): string => {
      if (ref.kind === "name") {
        const s = byName.get(ref.name);
        if (s === undefined)
          throw new Error(`schema ${from} references unknown schema ${ref.name}`);
        return s.name;
      }
      const s = byHash.get(ref.hash);
      if (s === undefined) {
        throw new Error(`schema ${from} references unknown pinned schema ${ref.hash} (E13)`);
      }
      return s.name;
    };
    const refs = new Map<string, string[]>();
    for (const s of schemas) {
      refs.set(
        s.name,
        collectRefs(s.body).map((r) => resolveName(r, s.name)),
      );
    }
    // DFS cycle detection over the resolved reference graph.
    const state = new Map<string, "visiting" | "done">();
    const visit = (name: string, path: string[]): void => {
      const st = state.get(name);
      if (st === "done") return;
      if (st === "visiting") {
        throw new Error(`schema reference cycle: ${[...path, name].join(" -> ")} (SPEC-3 §3)`);
      }
      state.set(name, "visiting");
      for (const r of refs.get(name) ?? []) visit(r, [...path, name]);
      state.set(name, "done");
    };
    for (const s of schemas) visit(s.name, []);
    return new SchemaRegistry(byName, byHash, readingsByName, readingsByHash);
  }

  get(name: string): HyperSchema | undefined {
    return this.byName.get(name);
  }

  getByHash(hash: string): HyperSchema | undefined {
    return this.byHash.get(hash);
  }

  resolve(ref: SchemaRefT): HyperSchema | undefined {
    return ref.kind === "name" ? this.byName.get(ref.name) : this.byHash.get(ref.hash);
  }

  getReading(name: string): Schema | undefined {
    return this.readingsByName.get(name);
  }

  resolveReading(ref: SchemaRefT): Schema | undefined {
    return ref.kind === "name"
      ? this.readingsByName.get(ref.name)
      : this.readingsByHash.get(ref.hash);
  }
}

File: implementations/rust/src/schema.rs
//! HyperSchemas and the schema registry (SPEC-3 §2-3 §6, ERRATA-2 E10/E13).
//! Mirrors ../ts/src/schema.ts: indexed by name AND by term hash; pinned refs resolve by hash.
//! Since issue #23 it also holds resolution Schemas ("readings"), indexed the same two ways, so
//! an expand term can name both halves of a child's lens and be validated at build time.

use std::collections::HashMap;

use crate::eval::{SchemaRef, Term};
use crate::resolution::Schema;
use crate::term_io::{schema_hash, term_hash};

#[derive(Debug, Clone, PartialEq)]
pub struct HyperSchema {
    pub name: String,
    /// L2 algebra version
    pub alg: u32,
    /// an HView-sort term, a function of the ambient root
    pub body: Term,
}

/// refs are derived from the body — every expand/fix schema reference (E10).
pub fn collect_refs(term: &Term) -> Vec<SchemaRef> {
    let mut out = Vec::new();
    fn walk(t: &Term, out: &mut Vec<SchemaRef>) {
        match t {
            Term::Input => {}
            Term::Select { of, .. }
            | Term::Mask { of, .. }
            | Term::Group { of, .. }
            | Term::Prune { of, .. }
            | Term::Resolve { of, .. } => walk(of, out),
            Term::Union { left, right } | Term::Intersect { left, right } => {
                walk(left, out);
                walk(right, out);
            }
            Term::Difference { of, without } => {
                walk(of, out);
                walk(without, out);
            }
            Term::Expand { schema, of, .. } => {
                out.push(schema.clone());
                walk(of, out);
            }
            Term::Fix { schema, .. } => out.push(schema.clone()),
        }
    }
    walk(term, &mut out);
    out
}

/// Reading refs are derived the same way — every expand's `reading` (issue #23). Kept separate
/// from collect_refs because they resolve against a different index (readings, not hyperschemas).
pub fn collect_reading_refs(term: &Term) -> Vec<SchemaRef> {
    let mut out = Vec::new();
    fn walk(t: &Term, out: &mut Vec<SchemaRef>) {
        match t {
            Term::Input | Term::Fix { .. } => {}
            Term::Select { of, .. }
            | Term::Mask { of, .. }
            | Term::Group { of, .. }
            | Term::Prune { of, .. }
            | Term::Resolve { of, .. } => walk(of, out),
            Term::Union { left, right } | Term::Intersect { left, right } => {
                walk(left, out);
                walk(right, out);
            }
            Term::Difference { of, without } => {
                walk(of, out);
                walk(without, out);
            }
            Term::Expand { reading, of, .. } => {
                if let Some(r) = reading {
                    out.push(r.clone());
                }
                walk(of, out);
            }
        }
    }
    walk(term, &mut out);
    out
}

#[derive(Debug, Clone, Default)]
pub struct SchemaRegistry {
    by_name: HashMap<String, HyperSchema>,
    by_hash: HashMap<String, HyperSchema>,
    readings_by_name: HashMap<String, Schema>,
    readings_by_hash: HashMap<String, Schema>,
}

impl SchemaRegistry {
    /// Rejects duplicate names, unresolved refs (gather AND reading), and reference cycles
    /// (SPEC-3 §3). Data cycles remain legal — the DAG constraint is on programs, not data.
    pub fn build(schemas: Vec<HyperSchema>, readings: Vec<Schema>) -> Result<Self, String> {
        let mut readings_by_name: HashMap<String, Schema> = HashMap::new();
        let mut readings_by_hash: HashMap<String, Schema> = HashMap::new();
        for r in &readings {
            let name = r
                .name
                .as_ref()
                .ok_or("a registered reading must carry a name (issue #23)")?;
            if readings_by_name.contains_key(name) {
                return Err(format!("duplicate reading name: {name}"));
            }
            readings_by_name.insert(name.clone(), r.clone());
            let h = schema_hash(r)?;
            // As with hyperschema bodies, two names MAY share a hash; first registration wins.
            readings_by_hash.entry(h).or_insert_with(|| r.clone());
        }
        let resolve_reading_ref = |r: &SchemaRef, from: &str| -> Result<(), String> {
            let found = match r {
                SchemaRef::Name(n) => readings_by_name.contains_key(n),
                SchemaRef::Pinned(h) => readings_by_hash.contains_key(h),
            };
            if found {
                return Ok(());
            }
            let label = match r {
                SchemaRef::Name(n) => n.clone(),
                SchemaRef::Pinned(h) => format!("pinned:{h}"),
            };
            Err(format!(
                "schema {from} references unknown reading {label} (issue #23)"
            ))
        };
        let mut by_name: HashMap<String, HyperSchema> = HashMap::new();
        let mut by_hash: HashMap<String, HyperSchema> = HashMap::new();
        for s in &schemas {
            if by_name.contains_key(&s.name) {
                return Err(format!("duplicate schema name: {}", s.name));
            }
            by_name.insert(s.name.clone(), s.clone());
            let h = term_hash(&s.body)?;
            // Two names MAY share a body hash; first registration wins the hash index.
            by_hash.entry(h).or_insert_with(|| s.clone());
            for r in collect_reading_refs(&s.body) {
                resolve_reading_ref(&r, &s.name)?;
            }
        }
        let resolve_name = |r: &SchemaRef, from: &str| -> Result<String, String> {
            match r {
                SchemaRef::Name(n) => by_name
                    .get(n)
                    .map(|s| s.name.clone())
                    .ok_or(format!("schema {from} references unknown schema {n}")),
                SchemaRef::Pinned(h) => by_hash.get(h).map(|s| s.name.clone()).ok_or(format!(
                    "schema {from} references unknown pinned schema {h} (E13)"
                )),
            }
        };
        let mut refs: HashMap<String, Vec<String>> = HashMap::new();
        for s in &schemas {
            let rs = collect_refs(&s.body)
                .iter()
                .map(|r| resolve_name(r, &s.name))
                .collect::<Result<Vec<_>, _>>()?;
            refs.insert(s.name.clone(), rs);
        }
        // DFS cycle detection over the resolved reference graph.
        #[derive(PartialEq)]
        enum State {
            Visiting,
            Done,
        }
        fn visit(
            name: &str,
            path: &mut Vec<String>,
            refs: &HashMap<String, Vec<String>>,
            state: &mut HashMap<String, State>,
        ) -> Result<(), String> {
            match state.get(name) {
                Some(State::Done) => return Ok(()),
                Some(State::Visiting) => {
                    return Err(format!(
                        "schema reference cycle: {} -> {name} (SPEC-3 §3)",
                        path.join(" -> ")
                    ));
                }
                None => {}
            }
            state.insert(name.to_string(), State::Visiting);
            path.push(name.to_string());
            if let Some(rs) = refs.get(name) {
                for r in rs {
                    visit(r, path, refs, state)?;
                }
            }
            path.pop();
            state.insert(name.to_string(), State::Done);
            Ok(())
        }
        let mut state = HashMap::new();
        for s in &schemas {
            visit(&s.name, &mut Vec::new(), &refs, &mut state)?;
        }
        Ok(Self {
            by_name,
            by_hash,
            readings_by_name,
            readings_by_hash,
        })
    }

    pub fn get(&self, name: &str) -> Option<&HyperSchema> {
        self.by_name.get(name)
    }

    pub fn get_by_hash(&self, hash: &str) -> Option<&HyperSchema> {
        self.by_hash.get(hash)
    }

    pub fn resolve(&self, r: &SchemaRef) -> Option<&HyperSchema> {
        match r {
            SchemaRef::Name(n) => self.by_name.get(n),
            SchemaRef::Pinned(h) => self.by_hash.get(h),
        }
    }

    pub fn get_reading(&self, name: &str) -> Option<&Schema> {
        self.readings_by_name.get(name)
    }

    pub fn resolve_reading(&self, r: &SchemaRef) -> Option<&Schema> {
        match r {
            SchemaRef::Name(n) => self.readings_by_name.get(n),
            SchemaRef::Pinned(h) => self.readings_by_hash.get(h),
        }
    }
}
